File: Core/src/main/java/jokatu/game/event/dialog/DialogRequestBuilder.java
package jokatu.game.event.dialog;

import jokatu.game.input.Input;
import jokatu.game.player.Player;
import ophelia.function.ExceptionalBiConsumer;
import org.jetbrains.annotations.NotNull;

public class DialogRequestBuilder<P extends Player> implements DialogRequestBuilderNeedingTitle {

	private final P player;

	private DialogRequestBuilder(@NotNull P player) {
		this.player = player;
	}

	@NotNull
	public static <P extends Player> DialogRequestBuilder<P> forPlayer(@NotNull P player) {
		return new DialogRequestBuilder<>(player);
	}

	@Override
	@NotNull
	public DialogRequestBuilderNeedingMessage<P> withTitle(@NotNull String title) {
		return new NeedingMessage(title);
	}

	private class NeedingMessage implements DialogRequestBuilderNeedingMessage<P> {
		private final String title;

		NeedingMessage(@NotNull String title) {
			this.title = title;
		}

		@NotNull
		@Override
		public DialogRequestBuilderNeedingInputType<P> withMessage(@NotNull String message) {
			return new NeedingInputType(message);
		}

		private class NeedingInputType implements DialogRequestBuilderNeedingInputType<P> {
			private final String message;

			NeedingInputType(@NotNull String message) {
				this.message = message;
			}

			@NotNull
			@Override
			public <I extends Input> DialogRequestBuilderNeedingConsumer<P, I> withInputType(@NotNull Class<I> inputType) {
				return new NeedingConsumer<>(inputType);
			}

			private class NeedingConsumer<I extends Input> implements DialogRequestBuilderNeedingConsumer<P, I> {
				private final Class<I> inputType;

				NeedingConsumer(@NotNull Class<I> inputType) {
					this.inputType = inputType;
				}

				@NotNull
				@Override
				public MainDialogRequestBuilder<P, I> withConsumer(@NotNull ExceptionalBiConsumer<? super I, ? super P, ?> consumer) {
					return new BaseDialogRequestBuilder<>(inputType, player, title, message, consumer);
				}
			}
		}
	}
}

File: Core/src/main/kotlin/jokatu/game/event/dialog/DialogRequestBuilder.kt
package jokatu.game.event.dialog

import jokatu.game.input.Input
import jokatu.game.player.Player
import ophelia.function.ExceptionalBiConsumer
import kotlin.reflect.KClass

/**
 * The beginning of sequence of builders that eventually result in a [DialogRequest] being sent to a player.
 * @param <P> the type of player to be presented with the dialog
 */
class DialogRequestBuilder<P : Player> private constructor(private val player: P) : DialogRequestBuilderNeedingTitle<P> {

	override fun withTitle(title: String): DialogRequestBuilderNeedingMessage<P> {
		return NeedingMessage(title)
	}

	private inner class NeedingMessage internal constructor(private val title: String) : DialogRequestBuilderNeedingMessage<P> {

		override fun withMessage(message: String): DialogRequestBuilderNeedingInputType<P> {
			return NeedingInputType(message)
		}

		private inner class NeedingInputType internal constructor(private val message: String) : DialogRequestBuilderNeedingInputType<P> {

			override fun <I : Input> withInputType(inputType: KClass<I>): DialogRequestBuilderNeedingConsumer<P, I> {
				return NeedingConsumer(inputType)
			}

			private inner class NeedingConsumer<I : Input> internal constructor(private val inputType: KClass<I>) : DialogRequestBuilderNeedingConsumer<P, I> {

				override fun withConsumer(consumer: ExceptionalBiConsumer<in I, in P, *>): MainDialogRequestBuilder<P, I> {
					return BaseDialogRequestBuilder(inputType, player, title, message, consumer)
				}
			}
		}
	}

	companion object {
		fun <P : Player> forPlayer(player: P): DialogRequestBuilder<P> {
			return DialogRequestBuilder(player)
		}
	}
}
